### ISSUE TITLE
`reload_config` causes systray (GDK) windows to die

### ISSUE THREAD

**Issue by ervinpopescu** [qtile/qtile#4897]
### Issue description

Example of `blueman-tray` output:
```bash
$ blueman-tray

(blueman-tray:58207): Gdk-WARNING **: 15:02:59.366: GdkWindow 0x2a00003 unexpectedly destroyed

(blueman-tray:58207): Gdk-WARNING **: 15:02:59.366: The program 'blueman-tray' received an X Window System error.
This probably reflects a bug in the program.
The error was 'BadWindow (invalid Window parameter)'.
  (Details: serial 281 error_code 3 request_code 18 (core protocol) minor_code 0)
  (Note to programmers: normally, X errors are reported asynchronously;
   that is, you will receive the error a while after causing it.
   To debug your program, run it with the GDK_SYNCHRONIZE environment
   variable to change this behavior. You can then get a meaningful
   backtrace from your debugger if you break on the gdk_x_error() function.)
```

Similar behavior with:
```
mictray
nm-applet
blueman-applet
systray_profile.py (custom GDK app)
pa-applet
```

With the default config:
```
$ blueman-tray

(blueman-tray:73413): Gdk-WARNING **: 15:13:19.501: GdkWindow 0x2a00002 unexpectedly destroyed

(blueman-tray:73413): Gdk-WARNING **: 15:13:19.502: The program 'blueman-tray' received an X Window System error.
This probably reflects a bug in the program.
The error was 'RenderBadPicture (invalid Picture parameter)'.
  (Details: serial 281 error_code 143 request_code 139 (RENDER) minor_code 7)
  (Note to programmers: normally, X errors are reported asynchronously;
   that is, you will receive the error a while after causing it.
   To debug your program, run it with the GDK_SYNCHRONIZE environment
   variable to change this behavior. You can then get a meaningful
   backtrace from your debugger if you break on the gdk_x_error() function.)
```

Not sure why the difference in errors. `RenderBadPicture (invalid Picture parameter)` vs `BadWindow (invalid Window parameter)`

`flameshot` (`qt`) is fine, so this is why I ruled out `qt`.

I would like some pointers for further debugging in addition to the current info returned by the error, if that's possible. Thanks in advance!

### Version

0.26.1.dev30+ge2158512

### Backend

X11 (default)

### Required

- [X] I have searched past issues to see if this bug has already been reported, and it hasn't been.
- [X] I understand that people give their precious time for free, and thus I've done my very best to make this problem as easy as possible to investigate.

**Comment by ervinpopescu** [comment]
Update: I bisected this to 7c3fbef756c451031e7d822dafed0b54eec0f2ac. @tych0 any idea why this happened and how we can fix it?

**Comment by ervinpopescu** [comment]
Is it just that `Systray` kills the window when [`Systray.finalize`](https://github.com/qtile/qtile/blob/master/libqtile/widget/systray.py#L278-L300) is called in [libqtile/bar.py:446](https://github.com/qtile/qtile/commit/7c3fbef756c451031e7d822dafed0b54eec0f2ac#diff-effdc1f9379ed88ce401e42f5852cc8db7444a9b885995477d68709069ca4977R446)?

**Comment by tych0** [comment]
Yeah, I think you're exactly right. So we need to do some magic there when we're reloading to propagate the window(s) in the systray across, but delete the drawers (or maybe propagate all of those too?). Oof :(

**Comment by ervinpopescu** [comment]
> Oof :(

:)

Why does `flameshot` (i.e. `qt`) not get killed?

**Comment by tych0** [comment]
Maybe it does, but it's smart enough to re-register? Not sure.

**Comment by tych0** [comment]
Hmm, this is weird. I wouldn't have expected a behavior change for non-WidgetBox using bars based on that patch. So clearly I don't understand something about how this is supposed to work :)

**Comment by ervinpopescu** [comment]
QQ: Have you managed to reproduce this?

**Comment by tych0** [comment]
I haven't had time to try yet, but I can definitely believe it's an issue. We should revert if it's an issue and we want to release, otherwise i'll try to track it down asap, probably next week.

**Comment by elParaguayo** [comment]
I don't think we need a new release yet so let's fix this if we can.

**Comment by ervinpopescu** [comment]
I'm fine with just running v0.26.0 and I don't see other people complaining, so I would say that a revert is not necessary as well.

Thanks for looking into it @tych0!

**Comment by jwijenbergh** [comment]
Fixed in git

**Comment by tych0** [comment]
Can we leave this one open? Ultimately we want to keep something like the patch so we don't leak widgets, but we need to find some way to persist them when we are restarting. I don't think it should be too hard to do, but I would like to have a reminder here :)

**Comment by jwijenbergh** [comment]
Ok no worries, sorry for closing!

**Comment by tych0** [comment]
All good, you are right that it is fixed right now :)

**Comment by ervinpopescu** [comment]
@tych0 reminder you wanted to keep this open as a reminder :)
